Sources/FoundationEssentials/Data/Data+Reading.swift

```diff
@@ -0,0 +1,345 @@
+    if options.contains(.uncached) {
+        // Non-zero arg turns off caching; we ignore error as uncached is just a hint
+        _ = fcntl(fd, F_NOCACHE, 1)
+    }
+#endif
+    
+    var filestat: stat = stat()
+    let err = fstat(fd, &filestat)
+    
+    guard err == 0 else {
+        throw fileReadingOrWritingError(posixErrno: errno, path: inPath, reading: true)
+    }
+    
+    // The following check is valid for 64-bit platforms.
+    if filestat.st_size > Int.max {
+        // We cannot hold this in `Data`, which uses Int as its count.
+        throw fileReadingOrWritingError(posixErrno: EFBIG, path: inPath, reading: true)
+    }
+    
+    let fileSize = min(Int(clamping: filestat.st_size), maxLength ?? Int.max)
+    let fileType = filestat.st_mode & S_IFMT
+    let preferredChunkSize = filestat.st_blksize
+#if !NO_FILESYSTEM
+    let shouldMap = shouldMapFileDescriptor(fd, path: inPath, options: options)
+#else
+    let shouldMap = false
+#endif
+        
+    if fileType != S_IFREG {
+        // EACCES is still an odd choice, but at least we have a better error for directories.
+        let code = (fileType == S_IFDIR) ? EISDIR : EACCES
+        throw fileReadingOrWritingError(posixErrno: code, path: inPath, reading: true)
+    }
+    
+    if fileSize < 0 {
+        throw fileReadingOrWritingError(posixErrno: ENOMEM, path: inPath, reading: true)
+    }
+    
+#if _pointerBitWidth(_32)
+    // Refuse to do more than 2 GB on 32-bit platforms
+    if fileSize > SSIZE_MAX {
+        throw fileReadingOrWritingError(posixErrno: EFBIG, path: inPath, reading: true)
+    }
+#endif
+    
+    let result: ReadBytesResult
+    let localProgress = (reportProgress && Progress.current() != nil) ? Progress(totalUnitCount: Int64(fileSize)) : nil
+    
+    if fileSize == 0 {
+        localProgress?.totalUnitCount = 1
+        localProgress?.completedUnitCount = 1
+        result = ReadBytesResult(bytes: nil, length: 0, deallocator: nil)
+    } else if shouldMap {
+#if !NO_FILESYSTEM
+        guard let bytes = mmap(nil, Int(fileSize), PROT_READ, MAP_PRIVATE, fd, 0) else {
+            throw fileReadingOrWritingError(posixErrno: errno, path: inPath, reading: true)
+        }
+        
+        guard bytes != MAP_FAILED else {
+            throw fileReadingOrWritingError(posixErrno: errno, path: inPath, reading: true)
+        }
+        
+        // Using bytes as the unit in this case doesn't really make any sense, since the amount of work required for mmap isn't meanginfully proportional to the size being mapped.
+        localProgress?.totalUnitCount = 1
+        localProgress?.completedUnitCount = 1
+        
+        result = ReadBytesResult(bytes: bytes, length: Int(fileSize), deallocator: .unmap)
+#else
+        // This was disabled above
+        fatalError("mapping should not be enabled")
+#endif
+    } else {
+        // We've verified above that fileSize will fit in `Int`
+        guard let bytes = malloc(Int(fileSize)) else {
```

Comment:
It's unfortunate that this always allocates. `Data` can store its contents [in an inline buffer](https://github.com/apple/swift-foundation/blob/87fdbaef20095a5c04dc952b291cc5f20d01bca6/Sources/FoundationEssentials/Data/Data.swift#L1265). 

It's not huge - on 64-bit systems, the inline buffer is 14 bytes, but for very small files it could be useful.

Reply:
I suspect the overhead of doing the I/O is much higher than a <=14 byte malloc, but your point is still valid. I think plumbing through the possibility of using inline storage would require some internal `Data` initializers, which we can follow up with.